src/api/accounts.rs

```diff
@@ -0,0 +1,515 @@
+                    Some(value) => Either::A(future::ok(value)),
+                    None => Either::B($f),
+                }
+            };
+        }
+
+        let from = key.public().address().into();
+        let inner = Future::join3(
+            maybe!(tx.nonce, accounts.web3().eth().transaction_count(from, None)),
+            maybe!(tx.gas_price, accounts.web3().eth().gas_price()),
+            maybe!(None, accounts.web3().net().version()),
+        );
+
+        SignTransactionFuture {
+            tx,
+            key: key.clone(),
+            inner,
+        }
+    }
+}
+
+impl<T: Transport> Future for SignTransactionFuture<T> {
+    type Item = SignedTransaction;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let (nonce, gas_price, chain_id) = try_ready!(self.inner.poll());
+
+        let chain_id = chain_id.parse::<u64>().map_err(|e| Error::Decoder(e.to_string()))?;
+        let tx = RawTxParameters {
+            nonce,
+            to: self.tx.to.unwrap_or_default(),
+            value: self.tx.value.unwrap_or_default(),
+            gas_price,
+            gas: self.tx.gas.unwrap_or(TRANSACTION_BASE_GAS_FEE),
+            data: self.tx.data.take().unwrap_or_default(),
+        };
+
+        let signed = tx.sign(&self.key, Some(chain_id))?;
+
+        Ok(Async::Ready(signed))
+    }
+}
+
+/// Trait for converting data into an `ethsign::Signature`.
+pub trait IntoSignature {
+    fn into_signature(self) -> Signature;
+}
+
+impl IntoSignature for (u8, H256, H256) {
+    fn into_signature(self) -> Signature {
+        let (v, r, s) = self;
+        Signature {
+            v: v - 27, // ¯\_(ツ)_/¯
+            r: r.into(),
+            s: s.into(),
+        }
+    }
+}
+
+impl<'a> IntoSignature for &'a Bytes {
+    fn into_signature(self) -> Signature {
+        if self.0.len() != 65 {
+            panic!("invalid signature bytes");
+        }
+
+        let v = self.0[64];
+        let r = H256::from_slice(&self.0[0..32]);
+        let s = H256::from_slice(&self.0[32..64]);
+
+        (v, r, s).into_signature()
+    }
+}
+
+impl<'a> IntoSignature for &'a SignedData {
+    fn into_signature(self) -> Signature {
+        (self.v, self.r, self.s).into_signature()
+    }
+}
+
+/// Raw transaction data for signing. When a transaction is signed, all
+/// parameter values need to be finalized for signing, and all parameters are
+/// required. Note that transaction data does not actually have the `from`
+/// public address as that is recoverable from the signature.
+#[derive(Debug)]
+struct RawTxParameters {
+    /// Transaction nonce (None for account transaction count)
+    nonce: U256,
+    /// To address
+    to: Address,
+    /// Supplied gas (None for sensible default)
+    gas: U256,
+    /// Gas price (None for sensible default)
+    gas_price: U256,
+    /// Transfered value (None for no transfer)
+    value: U256,
+    /// Data (None for empty data)
+    data: Bytes,
+}
```

Comment:
I've recently worked on a similar set of primitives:
https://github.com/tomusdrw/jsonrpc-proxy/blob/master/ethereum-proxy/plugins/accounts/transaction/src/lib.rs

Just published that to crates.io as `ethereum-transaction`. Feel free to re-use.

Reply:
This is great! I will change to use this instead.

Reply:
I just noticed that `ethereum-transaction` is using a different `ethereum-types` version which causes some issues. Are the plans on bumping `ethereum-types` version for `ethereum-transaction`?

Reply:
https://github.com/tomusdrw/jsonrpc-proxy/pull/93 Publishing as we speak.

Reply:
Awesome-sauce